File: ResourceManager/Services/TagDataService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Design;
using ResourceManager.Models;
using ResourceManager.Services.Abstractions;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Text;
using System.Threading.Tasks;

namespace ResourceManager.Services
{
    public class TagDataService : GenericDataService<Tag>, ITagDataService
    {
        public TagDataService(IDesignTimeDbContextFactory<ResourceManagerDBContext> contextFactory) : base(contextFactory)
        {
        }
        public override async Task<IList<Tag>> GetAll()
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                IList<Tag> entities = await db.Tags
                    .Include(x => x.PasswordTags)
                    .Include(x => x.RepositoryTags)
                    .Include(x => x.ActivityTags)
                    .Include(x => x.ArticleTags)
                    .Include(x => x.BookTags)
                    .AsNoTracking()
                    .OrderBy(x => x.Name)
                    .ToListAsync();
                return entities;
            }
        }
        public override async Task<IList<Tag>> GetWhere(Expression<Func<Tag, bool>> predicate)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                IList<Tag> entities = await db.Tags
                    .Include(x => x.PasswordTags)
                    .Include(x => x.RepositoryTags)
                    .Include(x => x.ActivityTags)
                    .Include(x => x.ArticleTags)
                    .Include(x => x.BookTags)
                    .Include(x => x.NoteTags)
                    .AsNoTracking()
                    .Where(predicate)
                    .OrderBy(x => x.Name)
                    .ToListAsync();
                return entities;
            }
        }
        public async Task<IList<Tag>> GetTagObjectForPasswordID(int id)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                var results = await (from t in db.Tags
                                     from ur in db.PasswordTags.Where(pt => pt.TagID == t.ID && pt.PasswordID == id)
                                     select t)
                                     .OrderBy(x => x.Name)
                                     .ToListAsync();
                return results;
            }
        }
        public async Task<IList<Tag>> GetTagObjectForBookID(int id)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                var results = await (from t in db.Tags
                                     from ur in db.BookTags.Where(pt => pt.TagID == t.ID && pt.BookID == id)
                                     select t)
                                     .OrderBy(x => x.Name)
                                     .ToListAsync();
                return results;
            }
        }
        public async Task<IList<Tag>> GetTagObjectForArticleID(int id)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                var results = await (from t in db.Tags
                                     from ur in db.ArticleTags.Where(pt => pt.TagID == t.ID && pt.ArticleID == id)
                                     select t)
                                     .OrderBy(x => x.Name)
                                     .ToListAsync();
                return results;
            }
        }
        public async Task<IList<Tag>> GetTagObjectForActivityID(int id)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                var results = await (from t in db.Tags
                                     from ur in db.ActivityTags.Where(pt => pt.TagID == t.ID && pt.ActivityID == id)
                                     select t)
                                     .OrderBy(x => x.Name)
                                     .ToListAsync();
                return results;
            }
        }
        public async Task<IList<Tag>> GetTagObjectForRepositoryID(int id)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                var results = await (from t in db.Tags
                                     from ur in db.RepositoryTags.Where(pt => pt.TagID == t.ID && pt.RepositoryID == id)
                                     select t)
                                     .OrderBy(x => x.Name)
                                     .ToListAsync();
                return results;
            }
        }
        public async Task<IList<Tag>> GetTagObjectForNoteID(int id)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                var results = await (from t in db.Tags
                                     from ur in db.NoteTags.Where(pt => pt.TagID == t.ID && pt.NoteID == id)
                                     select t)
                                     .OrderBy(x => x.Name)
                                     .ToListAsync();
                return results;
            }
        }
    }
}

File: ResourceManager/Services/TagDataService.vb
Imports System.Linq.Expressions
Imports Microsoft.EntityFrameworkCore
Imports Microsoft.EntityFrameworkCore.Design
Imports ResourceManager.Models
Imports ResourceManager.Services.Abstractions

Namespace ResourceManager.Services
    Public Class TagDataService
        Inherits GenericDataService(Of Tag)
        Implements ITagDataService
        Public Sub New(ByVal contextFactory As IDesignTimeDbContextFactory(Of ResourceManagerDBContext))
            MyBase.New(contextFactory)
        End Sub
        Public Overrides Async Function GetAll() As Task(Of IList(Of Tag)) Implements IDataService(Of Tag).GetAll
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim entities = Await db.Tags.Include(Function(x) x.PasswordTags) _
                    .Include(Function(x) x.RepositoryTags) _
                    .Include(Function(x) x.ActivityTags) _
                    .Include(Function(x) x.ArticleTags) _
                    .Include(Function(x) x.BookTags) _
                    .Include(Function(x) x.NoteTags) _
                    .AsNoTracking _
                    .OrderBy(Function(x) x.Name).ToListAsync
                Return entities
            End Using
        End Function
        Public Overrides Async Function GetWhere(ByVal predicate As Expression(Of Func(Of Tag, Boolean))) As Task(Of IList(Of Tag)) Implements IDataService(Of Tag).GetWhere
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim entities = Await db.Tags.Include(Function(x) x.PasswordTags) _
                    .Include(Function(x) x.RepositoryTags) _
                    .Include(Function(x) x.ActivityTags) _
                    .Include(Function(x) x.ArticleTags) _
                    .Include(Function(x) x.BookTags) _
                    .Include(Function(x) x.NoteTags) _
                    .AsNoTracking _
                    .Where(predicate) _
                    .OrderBy(Function(x) x.Name).ToListAsync
                Return entities
            End Using
        End Function
        Public Async Function GetTagObjectForPasswordID(ByVal id As Integer) As Task(Of IList(Of Tag)) Implements ITagDataService.GetTagObjectForPasswordID
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim results = Await (From t In db.Tags Join ur In db.PasswordTags
                                                           On t.ID Equals ur.TagID
                                     Where ur.PasswordID = id
                                     Order By t.Name
                                     Select t).ToListAsync
                Return results
            End Using
        End Function
        Public Async Function GetTagObjectForBookID(ByVal id As Integer) As Task(Of IList(Of Tag)) Implements ITagDataService.GetTagObjectForBookID
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim results = Await (From t In db.Tags Join ur In db.BookTags
                                                           On t.ID Equals ur.TagID
                                     Where ur.BookID = id
                                     Order By t.Name
                                     Select t).ToListAsync()
                Return results
            End Using
        End Function
        Public Async Function GetTagObjectForArticleID(ByVal id As Integer) As Task(Of IList(Of Tag)) Implements ITagDataService.GetTagObjectForArticleID
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim results = Await (From t In db.Tags Join ur In db.ArticleTags
                                                           On t.ID Equals ur.TagID
                                     Where ur.ArticleID = id
                                     Order By t.Name
                                     Select t).ToListAsync
                Return results
            End Using
        End Function
        Public Async Function GetTagObjectForActivityID(ByVal id As Integer) As Task(Of IList(Of Tag)) Implements ITagDataService.GetTagObjectForActivityID
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim results = Await (From t In db.Tags Join ur In db.ActivityTags
                                                           On t.ID Equals ur.TagID
                                     Where ur.ActivityID = id
                                     Order By t.Name
                                     Select t).ToListAsync
                Return results
            End Using
        End Function
        Public Async Function GetTagObjectForRepositoryID(ByVal id As Integer) As Task(Of IList(Of Tag)) Implements ITagDataService.GetTagObjectForRepositoryID
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim results = Await (From t In db.Tags Join ur In db.RepositoryTags
                                                           On t.ID Equals ur.TagID
                                     Where ur.RepositoryID = id
                                     Order By t.Name
                                     Select t).ToListAsync
                Return results
            End Using
        End Function
        Public Async Function GetTagObjectForNoteID(ByVal id As Integer) As Task(Of IList(Of Tag)) Implements ITagDataService.GetTagObjectForNoteID
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim results = Await (From t In db.Tags Join ur In db.NoteTags
                                                           On t.ID Equals ur.TagID
                                     Where ur.NoteID = id
                                     Order By t.Name
                                     Select t).ToListAsync
                Return results
            End Using
        End Function
    End Class
End Namespace
